internal/controller/providerconfig/backendstore_controller.go

```diff
@@ -61,6 +62,19 @@ func (r *BackendStoreReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	providerConfig := &apisv1alpha1.ProviderConfig{}
 	if err := r.kube.Get(ctx, req.NamespacedName, providerConfig); err != nil {
 		if kerrors.IsNotFound(err) {
+			// Delete the healthcheck bucket from the backend, before deleting the
+			// backend from the store. This is handled here instead of in the
+			// healthcheck-controller to avoid a race condition between
+			// deleting the bucket from the backend and deleting the backend from
+			// the store. Ideally this deletion would be handled when the bucket
+			// object is garbage collected, however the bucket controller's Delete
+			// method is not triggered by garbage collected buckets.
+			// TODO: Investigate this behaviour on garbage collection.
+			hcBucketName := req.Name + healthCheckSuffix
+			r.log.Info("Deleting health check bucket", "name", hcBucketName)
+			if err := s3internal.DeleteBucket(ctx, r.backendStore.GetBackend(req.Name), aws.String(hcBucketName)); err != nil {
+				return ctrl.Result{}, err
```

Comment:
Does it make sense to log and/or wrap this error? What if backend couldn't delete the health check? Backend in the store would be available for other go routines? Does it make sense to keep it in the store but mark as inactive, so we still can delete the health check, but we don't use the backend for new buckets?

Reply:
Going to merge this PR but tracking this comment here #23 to address in separate PR